wiring/src/spark_wiring_system.cpp

```diff
@@ -41,16 +47,101 @@ void SystemClass::reset(uint32_t data)
     HAL_Core_System_Reset_Ex(RESET_REASON_USER, data, nullptr);
 }
 
+SystemSleepResult SystemClass::sleep(const SystemSleepConfiguration& config) {
+    if (!config.valid()) {
+        LOG(ERROR, "System sleep configuration is invalid.");
+        System.systemSleepResult_ = SystemSleepResult(SYSTEM_ERROR_INVALID_ARGUMENT);
+    } else {
+        SystemSleepResult result;
+        int ret = system_sleep_ext(config.halConfig(), result.halWakeupSource(), nullptr);
+        result.setError(static_cast<system_error_t>(ret));
+        System.systemSleepResult_ = result;
+    }
+    System.toSleepResult();
+    return System.systemSleepResult_;
+}
+
 SleepResult SystemClass::sleep(Spark_Sleep_TypeDef sleepMode, long seconds, SleepOptionFlags flags)
 {
-    int ret = system_sleep(sleepMode, seconds, flags.value(), NULL);
-    System.sleepResult_ = SleepResult(WAKEUP_REASON_NONE, static_cast<system_error_t>(ret));
-    return System.sleepResult_;
+    if (sleepMode == SLEEP_MODE_WLAN) {
```

Comment:
Let's keep the compatibility layer in the system instead. We anyway need to keep both system and HAL sleep 1.0 functions around for ABI compatibility, so it does not make sense to have this in wiring.

Reply:
Done.